packages/next/src/server/app-render/action-encryption-utils.ts

```diff
@@ -0,0 +1,195 @@
+import type { ActionManifest } from '../../build/webpack/plugins/flight-client-entry-plugin'
+import type { ClientReferenceManifest } from '../../build/webpack/plugins/flight-manifest-plugin'
+
+// Keep the key in memory as it should never change during the lifetime of the server in
+// both development and production.
+let __next_loaded_action_key: CryptoKey
+let __next_internal_development_raw_action_key: string
+
+export function arrayBufferToString(buffer: ArrayBuffer) {
+  const bytes = new Uint8Array(buffer)
+  const len = bytes.byteLength
+
+  // @anonrig: V8 has a limit of 65535 arguments in a function.
+  // For len < 65535, this is faster.
+  // https://github.com/vercel/next.js/pull/56377#pullrequestreview-1656181623
+  if (len < 65535) {
+    return String.fromCharCode.apply(null, bytes as unknown as number[])
+  }
+
+  let binary = ''
+  for (let i = 0; i < len; i++) {
+    binary += String.fromCharCode(bytes[i])
+  }
+  return binary
+}
+
+export function stringToUint8Array(binary: string) {
+  const len = binary.length
+  const arr = new Uint8Array(len)
+
+  for (let i = 0; i < len; i++) {
+    arr[i] = binary.charCodeAt(i)
+  }
+
+  return arr
+}
+
+export async function encrypt(key: CryptoKey, salt: string, data: Uint8Array) {
+  const iv = new TextEncoder().encode(salt)
```

Comment:
Is there a reason for nor reusing the text encoders?

Reply:
Nope, will make it reusable! Thanks